### DISCUSSION TITLE
Code span in hyperref URL

### DISCUSSION THREAD

**Discussion by csrmy** [Witiko/markdown#493] [Q&A] [answered]
Hi,
I am new to TeX and markdown.

I want to be able to style an hyperlink in a code span like so:
Lorem ipsum [`dolor sit`](https://en.wiktionary.org/wiki/dolor) amet

The actual render I have:
![image](https://github.com/user-attachments/assets/3b1ea75c-7df7-4fb7-9aff-ff0a0339d40c)

**The link works, and redirect correctly**

TeX File:
```
\documentclass{article}

\usepackage{hyperref} % package to use hypertext references
\usepackage{markdown} % package to parse markdown in Tex
\usepackage{verbatim} % package to sanatize text
\usepackage{tcolorbox} % package to add styled boxes

% style for hypertext
\hypersetup{
	colorlinks=true,
	linkcolor=blue,
	filecolor=cyan,
	urlcolor=cyan
}
\urlstyle{same}

% markdown settings
\markdownSetup{%
	outputDir=_build-geotex, % set output directory
	fencedCode=true, % allow code blocks
	fancyLists=true, % allow more ways to create lists
	texMathDollars=true, % allow Tex dollar sign mathematical expression
	notes=true,	% allow bottom page notes
	pipeTables=true, % allow markdown tables
	taskLists=true, % allow task lists
	underscores=true, % remove usage of underscores for emphasis
	subscripts=true, % allow small bottom characters
	superscripts=true % allow small top characters
}

% define inline code style
\newtcbox{\codeinline}[1][]{
verbatim,
on line,
arc=0mm,
boxsep=0mm,
left=3pt,
right=3pt,
top=1pt,
bottom=1pt,
boxrule=1pt}

% override markdown render of inline code
\global\def\markdownRendererCodeSpan#1{%
	\codeinline{\texttt{#1}}
}

% override hyperlinks to render inline
\global\def\markdownRendererLink#1#2#3#4{%
  \href{#3}{#1}
}

\begin{document}
\markdownInput{test.md}
\end{document}
```

markdown file:
```
Regular link: [lien](https://github.com/Witiko/markdown)

Link inline code: [`lien`](https://github.com/Witiko/markdown)
```

Thanks for your help and time.

**Comment by Witiko** [comment]
With your definitions, ```[`lien`](https://github.com/Witiko/markdown)``` becomes `\href{https://github.com/Witiko/markdown}{\codeinline{\texttt{lien}}}`. Apparently, `\codeinline` discards the blue text color from `\href`.

Perhaps you can set up a conditional such as `\ifinsidehyperlink`, which you would set to true at the beginning of `\markdownRendererLink` and to false at the end. Then, you could use `\textcolor{blue}{#1}` inside `\codeinline{\texttt{...}}` in your definition of `\markdownRendererCodeSpan` when `\ifinsidehyperlink` is true.

**Reply by csrmy** [reply]
I don't know much on logical programming on Tex or LaTex. @Witiko Is there a better way for conditional statements ?

I also searched to use `\tl_if_eq` but I could not figure how to make it work.

I came up with this, probably not the best way at all :
``` tex
\newcommand{\islink}{FALSE}

% override markdown render of inline code
\global\def\markdownRendererCodeSpan#1{%
	\codeinline{%
		\ifthenelse{\equal{\islink}{TRUE}}%
		{%
			\textcolor{cyan}{#1}%
		}
		{%
			#1%
		}
	}
}

% override hyperlinks to render inline
\global\def\markdownRendererLink#1#2#3#4{%
	\renewcommand{\islink}{TRUE}
	\href{#3}{#1}
	\renewcommand{\islink}{FALSE}
}
```

markdown:
```
Regular link: [link](https://github.com/Witiko/markdown)

Link inline code: [`link` still link but no code](https://github.com/Witiko/markdown)

Inline code: `ls -la`
```

Result is good :
![image](https://github.com/user-attachments/assets/c3884302-6a64-4501-86c3-906761da0fb9)

**Reply by Witiko** [reply]
Looks good!

In expl3, you might do something like this:

``` tex
\ExplSyntaxOn

\bool_new:N \l_islink_bool  % create a new boolean variable

% override markdown render of inline code
\cs_gset:Npn
  \markdownRendererCodeSpan#1
  {
    \codeinline
      {
        \bool_if:NTF
          \l_islink_bool
          { \textcolor { cyan } { #1 } }
          { #1 }
      }
  }

% override hyperlinks to render inline
\cs_gset:Npn
  \markdownRendererLink#1#2#3#4
  {
    \bool_set_true:N
      \l_islink_bool
    \href { #3 } { #1 }
    \bool_set_false:N
      \l_islink_bool
  }

\ExplSyntaxOff
```

**Reply by csrmy** [reply]
@Witiko Thank you for taking the time to provide expl3 syntax, I want to learn more about it (it's very tough to grasp at first).

I tried to import your code but got this `! LaTeX Error: Control sequence \markdownRendererCodeSpan already defined.`

Is it expected ?

**Reply by Witiko** [reply]
`\cs_new:Npn` will refuse to redefine an existing macro. Can you try `\cs_gset:Npn` instead?

Sorry about this: I didn't test the code, I am on mobile.

**Comment by csrmy** [comment, marked as answer]
Based on @Witiko 's answers.

The issue is to style the à `code span` when it's part of a hyperref URL.

**Markdown source**:
```
[link `also link`]([google.com](https://github.com/Witiko/markdown))
```

## Origin
**Origin render**:
> Not the default render of link or code span

![image](https://github.com/user-attachments/assets/e5887291-1030-41f7-970b-1779fc18bc70)

**Origin Tex**:
``` tex
% override hyperlinks to render inline
\global\def\markdownRendererLink#1#2#3#4{%
  \href{#3}{#1}%
}

% override markdown render of inline code
\global\def\markdownRendererCodeSpan#1{%
	\codeinline{#1}
}
```

## Fix
To be able to style a code span depending on if it's a link or not, we need to save if a `markdownRendererLink` is being rendered with the help of a variable. Then when rendering a code span with `markdownRendererCodeSpan`, we can check this value and style it like a link or not.

### expl3 syntax
``` latex
\ExplSyntaxOn% enable expl3 syntax

% declare boolean value named: isRenderingHyperLink
\bool_new:N \l_isRenderingHyperLink_bool

% override markdown render of inline code
\cs_gset:Npn
	\markdownRendererCodeSpan#1
	{
		\codeinline
		{
			% if value is true color text in cyan
			\bool_if:NTF \l_isRenderingHyperLink_bool
			{ \textcolor { cyan } { #1 } }
			{ #1 }
		}
	}

% override hyperlinks to render inline
\cs_gset:Npn
	\markdownRendererLink#1#2#3#4
	{
		% set boolean value to true
		\bool_set_true:N \l_isRenderingHyperLink_bool
		\href { #3 } { #1 }
		% set boolean value to false
		\bool_set_false:N \l_isRenderingHyperLink_bool
	}

\ExplSyntaxOff% disable expl3 syntax
```

### ifthen package syntax
``` latex
% declare boolean value named: islink
\newcommand{\islink}{FALSE}

% override markdown render of inline code
\global\def\markdownRendererCodeSpan#1{%
	\codeinline{%
		\ifthenelse{\equal{\islink}{TRUE}}%
		{%
			\textcolor{cyan}{#1}%
		}
		{%
			#1%
		}
	}
}

% override hyperlinks to render inline
\global\def\markdownRendererLink#1#2#3#4{%
	\renewcommand{\islink}{TRUE}% set boolean value to true
	\href{#3}{#1}
	\renewcommand{\islink}{FALSE}% set boolean value to false
}
```

**In both approach, the result is**:
![image](https://github.com/user-attachments/assets/d930d602-3995-4443-99c2-10fc9b4c395c)

The `\codeinline` declaration, using the [`tcolorbox`](https://ctan.org/pkg/tcolorbox) package:
``` latex
% define inline code style
\newtcbox{\codeinline}[1][]{
verbatim,
on line,
arc=0mm,
boxsep=0mm,
left=3pt,
right=3pt,
top=1pt,
bottom=1pt,
boxrule=1pt}
```

Thank you @Witiko for your time and work.
